### DISCUSSION TITLE
How to store the maximum value in each cell at all moments during the calculation?

### DISCUSSION THREAD

**Discussion by MaybeBin** [AMReX-Codes/amrex#4209] [Q&A] [answered]
Hi ~
I used the CNS sample code as the basis for my implementation and extended it by adding an additional source term for combustion calculations. My goal is to simulate detonation and identify the detonation cell by recording the history of the maximum pressure in each cell, as shown in the figure.
![56889d517aedab0c39e809e421a1b80](https://github.com/user-attachments/assets/c0fc6683-b93c-46f5-a84c-440dd2aca0d5)
I attempted to implement this by defining a derived variable function as shown below, but I noticed that the values in `derfab` are reinitialized after each time step, preventing them from storing the maximum value across all time steps. I’m not certain if my understanding is correct. Could you please provide any advice on how to implement this on an AMR grid?
![image](https://github.com/user-attachments/assets/96e8e985-dd8e-46ae-adac-3ca9d35efd5d)

Many thanks!

**Comment by zingale** [comment, marked as answer]
I would go about this by making an additional StateData with just a single time-level and single component, and then at the end of the evolution loop, do the max of the current pressure and what's stored there.  This way you can have it stored in the checkpoint for you, so it restarts.  

(Also, fun bit -- we did cell dets ~ 20 years ago, but I made that plot from the plotfiles back then: https://iopscience.iop.org/article/10.1086/317135)

**Reply by MaybeBin** [reply]
Thanks for your advice and I'll try it :)

**Comment by zingale** [comment]
one further thought -- if you do this, then that max pressure state will regrid along with the simulation, which perhaps is not what you want, especially if you concentrate the refinement on the detonation front itself.  That's why it might be better to just output pressure into plotfiles frequently (maybe as small plotfiles) and then use a postprocessing tool to create a uniform domain and do a zone-by-zone max over the time-series of plotfiles.

**Reply by himcraft** [reply]
Hi Dr. Zingale, thank you for your advice above. Is it possible to manually allocate a StateData at the finest possible mesh level to store these maximals? So we won't need to worry about regridding.

**Reply by zingale** [reply]
I am not sure.   I think you would need a separate DistributionMapping for it.

**Reply by himcraft** [reply]
Thank you very much! I will look into it.
